### PR TITLE
Rect and FRect swizzling

### PR DESCRIPTION
**Pull request by ScriptLineStudios**
Swizzling for Rects and FRects similar to Vector and Color swizzling

### CODE REVIEW

File: test/rect_test.py [hunk 1]
```diff
@@ -2691,6 +2728,50 @@ def test_contains(self):
...
+        self.assertAlmostEqual(res[0], 10.34, 5)
+        self.assertAlmostEqual(res[1], 10.34, 5)
+
+        res = c.xyw
+        self.assertAlmostEqual(res[0], 10.34, 5)
+        self.assertAlmostEqual(res[1], 20.31, 5)
+        self.assertAlmostEqual(res[2], 30.39, 5)
+
+        with self.assertRaises(AttributeError):
+            c.rr = (10.3421784, 10.3421784)
```
Comment (oddbookworm) [inline test/rect_test.py:2752]: Another copy/paste error?

File: test/rect_test.py [hunk 2]
```diff
@@ -2334,6 +2334,43 @@ def test_collideobjectsall_call_variants(self):
...
+        c.xywh = c.hwyx
+        self.assertEqual(c.xywh, (40, 30, 20, 10))
+
+        with self.assertRaises(AttributeError):
+            c.rr = (10, 10)
+            c.bb = (10, 10)
+            c.gg = (10, 10)
+            c.aa = (10, 10)
+            c.aaa = (10, 10, 100)
+            c.rgb = (256, 256, 256)
```
Comment (oddbookworm) [inline test/rect_test.py:2355]: Copy/Paste error? `c` is a `Rect`, so it doesn't have these properties in the first place

File: docs/reST/ref/rect.rst [hunk 3]
```diff
@@ -32,6 +32,13 @@
    None but affects the original Rect. These "in-place" methods are denoted
    with the "ip" suffix.
 
+   Rect objects support swizzling for their ``XYWH`` attributes, which allows
+   the creation of new color objects with the corresponding swizzled attributes
```
Comment (oddbookworm) [inline docs/reST/ref/rect.rst:36]: copy/paste error? "color"

File: src_c/rect.c [hunk 4]
```diff
@@ -438,6 +442,402 @@ four_floats_from_obj(PyObject *obj, float *val1, float *val2, float *val3,
...
+    attr = PyUnicode_AsUTF8AndSize(attr_unicode, &len);
+    if (attr == NULL)
+        goto internal_error;
+    /* If we are not a swizzle, go straight to GenericGetAttr. */
+    if ((attr[0] != 'x') && (attr[0] != 'y') && (attr[0] != 'w') &&
+        (attr[0] != 'h')) {
+        goto swizzle_failed;
+    }
+
+    if (len == 4) {
```
Comment (oddbookworm) [inline src_c/rect.c:474]: Like with swizzling colors, I'm not a fan of having a different return type in special cases. I think it should just return a tuple even when 4 values are requested

**Review by oddbookworm — CHANGES_REQUESTED** [review]

File: src_c/rect.c (hunk 4 above)
Comment (Matiiss) [reply src_c/rect.c]: For `Vector3` swizzling two values returns a `Vector2` though and three values return a `Vector3`, also despite them being different types they can both be iterated over, the only loss here is immutability (but Vectors have it very similar).

File: test/rect_test.py [hunk 5]
```diff
@@ -2334,6 +2334,36 @@ def test_collideobjectsall_call_variants(self):
         r.collideobjectsall(objects, key=lambda o: o.rect1)
         self.assertRaises(TypeError, r.collideobjectsall, objects)
 
+    def test_swizzle_get(self):
```
Comment (Starbuck5) [inline test/rect_test.py:2337]: No test that a 4 element swizzle returns a Rect?

**Comment by Starbuck5** [comment]
I'm assuming you got most of the implementation by checking the vector code, that should be good. Before adding this, I'd like to hear a use case for why this should be added.

**Comment by MyreMylar** [comment]
> I'm assuming you got most of the implementation by checking the vector code, that should be good. Before merging this, I'd like to hear a use case for why this should be added.

I'm also struggling to think of a reason I would use this. I guess flipping coordinates around with rectangles is just less common in game development? 

If anybody out there has a use case for this then I am totally open to hearing it.

**Comment by Matiiss** [comment]
The only somewhat useful case I can think of would be if the swizzling returned a `pygame.Vector2` when swizzling two values. That way you could get the center of the rect as a vector like so:
```py
center_vec = rect.xy + rect.wh / 2
```
compared to
```py
center_vec = pygame.Vector2(rect.center)
```
So yeah... not a very strong argument, especially given that we could add `_v` attributes to the rects, like `center_v` that would achieve the same thing

**Comment by MyreMylar** [comment]
No clear uses cases for this, seem to have emerged. In a search I can't find any other game or graphic library that has added swizzling to geometry other than vectors I'm going to move that we close this PR. 

Thankyou for the work put into this @ScriptLineStudios, but I think it is not a good idea for us to add new functions that it seems like nobody will use as it just creates more code to maintain in the future and complicates the public API for pygame.

If I'm wrong, please re-open this with a compelling use case or demand for this feature and I'd be happy for it to go in.

**Pull request closed without merging**